bindings/rust-examples/async-pkey-offload/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+        .key_spec(KeySpec::EccNistP384)
+        .key_usage(aws_sdk_kms::types::KeyUsageType::SignVerify)
+        .description(KEY_DESCRIPTION)
+        .send()
+        .await?;
+    println!("creating new key");
+    let key = create_key_resp.key_metadata.unwrap().key_id;
+    Ok(key)
+}
+
+/// KmsAsymmetricKey is a container used to implement application-specific traits.
+///
+/// It implements [rcgen::RemoteKeyPair] which allows us to create a self-signed
+/// x509 cert corresponding to the key pair.
+///
+/// It implements [s2n_tls::callbacks::PrivateKeyCallback] which allows us to offload
+/// cryptographic operations from s2n-tls to the KMS key.
+#[derive(Debug, Clone)]
+pub struct KmsAsymmetricKey {
+    /// AWS KMS SDK client.
+    kms_client: Client,
+    /// A copy of the public key in "raw" format
+    public_key: Vec<u8>,
+    /// The KMS key id
+    key_id: String,
+}
+
+impl KmsAsymmetricKey {
+    const EXPECTED_SIG: s2n_tls::enums::SignatureAlgorithm =
+        s2n_tls::enums::SignatureAlgorithm::ECDSA;
+
+    /// Encapsulate an existing KmsAsymmetricKey
+    ///
+    /// This method does not create a new key in KMS. It will retrieve the public
+    /// key of an existing key to be used locally.
+    pub async fn new(client: Client, key_id: String) -> Result<Self, Box<dyn std::error::Error>> {
+        let public_key_output = client
+            .get_public_key()
+            .key_id(key_id.clone())
+            .send()
+            .await?;
+        // > The public key that AWS KMS returns is a DER-encoded X.509 public key,
+        // > also known as SubjectPublicKeyInfo (SPKI), as defined in RFC 5280.
+        // > When you use the HTTP API or the AWS CLI, the value is Base64-encoded.
+        // > Otherwise, it is not Base64-encoded.
+        // https://docs.aws.amazon.com/kms/latest/developerguide/download-public-key.html
+        // Note that the rust sdk seems to handle common encoding tasks for
+        // us, so `encoded_public_key` is binary, not base64 encoded.
+        let encoded_public_key = public_key_output.public_key.unwrap().into_inner();
+        let raw_public_key = extract_ex_public_key(&encoded_public_key)?;
+
+        Ok(Self {
+            kms_client: client,
+            public_key: raw_public_key,
+            key_id,
+        })
+    }
+
+    /// Perform an async pkey offload.
+    ///
+    /// s2n-tls requires that future have 'static bounds, so this function can not
+    /// operation on `&self`. Instead we clone all of the necessary elements and
+    /// capture them in the closure.
+    async fn async_pkey_offload(
+        kms_client: Client,
+        key_id: String,
+        operation: s2n_tls::callbacks::PrivateKeyOperation,
+        kms_key_spec: SigningAlgorithmSpec,
+    ) -> Result<(PrivateKeyOperation, Vec<u8>), s2n_tls::error::Error> {
+        //> If this is an OperationType::Sign operation, then this input has
+        //> already been hashed and is the resultant digest.
```

Comment:
If this is a quote from somewhere, it needs attribution. If it's not a quote, idk what the ">" are about.